### DISCUSSION TITLE
How to test Entities with Collections? Missing docs?

### DISCUSSION THREAD

**Discussion by DavidGarciaCat** [mikro-orm/mikro-orm#3605] [Q&A] [answered]
Good day,

I am building an API with TypeScript and ExpressJS, and I used Mikro ORM to map my entities. The API itself works as expected, and the entities are stored and mapped with their primary keys, so there are no problems at this end.

However, I am unable to implement Jest tests for my entities (as they have more than just getters, setters and property mapping with the decorators) and my Jest tests are throwing the following error:

```
TypeError: Cannot read properties of undefined (reading '__meta')
```

I have noticed this is happening when my entity has a Collection, as Jest points the error is caused inside my constructor, for cases like this:

```ts
// scenario 1: one user with many profiles
// user constructor has...
this.profiles.add(new Profile( this, '...' ));

// scenario 2: a wallet can work with many currencies
// wallet constructor has...
this.currencies.add(new Currency( this, 'EUR' ));
```

I have checked the "Real-World Example App" for guidance, but I can see there are no tests for the entities:
https://github.com/mikro-orm/nestjs-realworld-example-app/tree/master/src/user/test

According to this Stack Overflow page, a potential solution would be running the `MikroORM.init( { ... } )` and setting the 2nd parameter (so entities are read and mapped but the connection doesn't initialise). I have also tried it, but I still get the same error message.
https://stackoverflow.com/questions/69924546/unit-testing-mirkoorm-entities

I couldn't find an example, so I would like to ask for it, please, and also an update of the docs so this can be available for other devs strugling like me.

Thank you,

**Comment by B4nan** [comment, marked as answer]
As the SO post says (= as I am saying there on SO), you need to initialize the ORM to be able to work with entities - the discovery needs to happen, otherwise, the [propagation](https://mikro-orm.io/docs/propagation) won't work. 

I have a new getting started guide in progress, it will cover the basics of testing too, but it will take some time to finish.

>  I have also tried it, but I still get the same error message.

Provide a repro then, I'd guess you initialize the ORM too late. Also provide full stack trace, not just the error message, that's not really helpful.

**Reply by DavidGarciaCat** [reply]
Funny thing, as I didn't notice it was you :-D

Unfortunately, this project is hosted in a private repo, so I can't share the link per-se.
But, please, give me a few minutes to make a replica, so I can share it with you.

In the meantime, is there any web link where this "getting started guide" is posted with an example, even if it's not 100% ready?

**Reply by B4nan** [reply]
Nope, its a private repo currently, and the part about testing it not there yet. But its really all said in that SO post, if you want to test without touching a database, you still need to init the ORM. You can opt out of connecitng to the database - in latest version you can also just pass `connect: false` in the ORM config, no need for the second parameter.

Note that I want to see a reproduction, not your project. It needs to be minimal.

**Reply by DavidGarciaCat** [reply]
Yes, I know - that's why I asked for a few minutes.

So, here there is a minimal project: https://github.com/DavidGarciaCat/mikro-orm-example

In case you need it, there is a `docker-compose.yml` file to boot a test database

if you run the `npm test` command, then you should see this:

```
 FAIL  tests/model/user.test.ts
  User model
    ✕ exposes the email (1 ms)
    ✕ exposes the number of profiles

  ● User model › exposes the email

    TypeError: Cannot read properties of undefined (reading '__meta')

      19 |
      20 |         this.email = email;
    > 21 |         this.profiles.add(profile);
         |                       ^
      22 |     }
      23 | }
      24 |

      at Collection.get property [as property] (node_modules/@mikro-orm/core/entity/ArrayCollection.js:151:56)
      at Collection.validateItemType (node_modules/@mikro-orm/core/entity/Collection.js:313:71)
      at node_modules/@mikro-orm/core/entity/Collection.js:95:40
          at Array.forEach (<anonymous>)
      at Collection.add (node_modules/@mikro-orm/core/entity/Collection.js:95:19)
      at new add (src/model/user.ts:21:23)
      at Object.<anonymous> (tests/model/user.test.ts:7:22)

  ● User model › exposes the number of profiles

    TypeError: Cannot read properties of undefined (reading '__meta')

      19 |
      20 |         this.email = email;
    > 21 |         this.profiles.add(profile);
         |                       ^
      22 |     }
      23 | }
      24 |

      at Collection.get property [as property] (node_modules/@mikro-orm/core/entity/ArrayCollection.js:151:56)
      at Collection.validateItemType (node_modules/@mikro-orm/core/entity/Collection.js:313:71)
      at node_modules/@mikro-orm/core/entity/Collection.js:95:40
          at Array.forEach (<anonymous>)
      at Collection.add (node_modules/@mikro-orm/core/entity/Collection.js:95:19)
      at new add (src/model/user.ts:21:23)
      at Object.<anonymous> (tests/model/user.test.ts:12:22)
```

**Reply by B4nan** [reply]
Well, your test is not initializing the ORM before it runs. You cant just wrap async code in IIFE and expect it to finish before resolving the module. You need to await things, you dont do that, you just run the async IIFE and work with the static `ORM` export before it resolves.

**Reply by DavidGarciaCat** [reply]
what approach do you think I should apply here to make sure it's being initialised? just run any `ORM.orm...` call with an `async` inside the `beforeEach( () => {} )` method?

If you think there is a better approach to solve this problem, ether in the tests files or as part of the config, please let me know

**Reply by B4nan** [reply]
Check the tests in here, they all use the same approach as you need - initialize the ORM in `beforeAll` and close it in `afterAll`.

https://github.com/mikro-orm/mikro-orm/tree/master/tests/issues

**Reply by DavidGarciaCat** [reply]
I have updated the same repo I shared above.

Unless I made a mistake, now I am taking a new instance of the ORM by exporting and importing the result of `MikroORM.imit()`. Also, it seems the code coverage report now returns green lines everywhere:

<img width="799" alt="Screenshot 2022-10-18 at 20 24 29" src="https://user-images.githubusercontent.com/7809429/196513791-e516ab46-cd00-4004-b499-841a3f25b441.png">

Still... it looks like the process to close the connection doesn't work - I assume this is because if we pass the 2nd argument to tell the ORM client **not to connect** then there's no connection to close.

----

All this leads me to the next question:

If I **don't want** to establish a real DB connection for unit tests, but I **do** for integration tests, then I assume I should consider using a function/class and pass an argument to specify when I want to connect and when I don't want to, correct?

**Reply by DavidGarciaCat** [reply]
I have updated the repo once more to include ⬆️ this approach (and it seems to work for my needs).
Still, if you think there is a better way to proceed, I am more than happy to discuss it.

https://github.com/DavidGarciaCat/mikro-orm-example

Many thanks for pointing out the example repo, it helped a lot to understand the issue.

**Reply by B4nan** [reply]
> Still... it looks like the process to close the connection doesn't work - I assume this is because if we pass the 2nd argument to tell the ORM client not to connect then there's no connection to close.

indeed, if you dont open the connection, you dont need to close it. i guess we should allow calling `orm.close()` in that case instead of failing, there's indeed room for improvement

**Reply by DavidGarciaCat** [reply]
Both improving and learning never end 😄
